File: app/src/main/java/website/todds/saundio/components/MediaService.java
package website.todds.saundio.components;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.media.MediaBrowserCompat;
import android.support.v4.media.MediaBrowserServiceCompat;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.text.TextUtils;

import java.util.List;

import website.todds.saundio.R;

import static android.support.v4.media.session.PlaybackStateCompat.*;

public class MediaService extends MediaBrowserServiceCompat {

    private MediaSessionCompat mMediaSession;
    private PlaybackStateCompat.Builder mStateBuilder;
    private SaundioMediaCallback mCallbacks;

    private static final long POSSIBLE_STATE_FLAGS =
            ACTION_PLAY |
            ACTION_PAUSE |
            ACTION_PLAY_PAUSE |
            ACTION_PLAY_FROM_MEDIA_ID |
            ACTION_PLAY_FROM_URI |
            ACTION_PLAY_FROM_SEARCH |
            ACTION_FAST_FORWARD |
            ACTION_REWIND |
            ACTION_SKIP_TO_NEXT |
            ACTION_SKIP_TO_PREVIOUS |
            ACTION_SEEK_TO;

    @Override
    public void onCreate() {
        super.onCreate();

        mMediaSession = new MediaSessionCompat(getApplicationContext(), MediaService.class.getSimpleName());
        mMediaSession.setFlags(MediaSessionCompat.FLAG_HANDLES_MEDIA_BUTTONS |
                MediaSessionCompat.FLAG_HANDLES_TRANSPORT_CONTROLS);

        mStateBuilder = new PlaybackStateCompat.Builder().setActions(POSSIBLE_STATE_FLAGS);
        mMediaSession.setPlaybackState(mStateBuilder.build());

        mCallbacks = new SaundioMediaCallback(this, mMediaSession, mStateBuilder);
        mMediaSession.setCallback(mCallbacks);

        setSessionToken(mMediaSession.getSessionToken());
    }

    @Override
    public void onDestroy() {
        if (mMediaSession != null) {
            mMediaSession.release();
            mMediaSession = null;
        }
        mStateBuilder = null;
        mCallbacks = null;
        super.onDestroy();
    }

    @Nullable
    @Override
    public BrowserRoot onGetRoot(@NonNull String clientPackageName, int clientUid, @Nullable Bundle rootHints) {
        if (TextUtils.equals(clientPackageName, getPackageName()))
            return new BrowserRoot(getString(R.string.app_name), null);

        return null;
    }

    @Override
    public void onLoadChildren(@NonNull String parentId, @NonNull Result<List<MediaBrowserCompat.MediaItem>> result) {
        result.sendResult(null);
    }
}

File: app/src/main/kotlin/website/todds/saundio/components/MediaService.kt
package website.todds.saundio.components

import android.os.Bundle
import android.support.v4.media.MediaBrowserCompat
import android.support.v4.media.MediaBrowserServiceCompat
import android.support.v4.media.session.MediaSessionCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.text.TextUtils

import website.todds.saundio.R

import android.support.v4.media.session.PlaybackStateCompat.*

class MediaService : MediaBrowserServiceCompat() {

    private lateinit var mMediaSession: MediaSessionCompat
    private lateinit var mStateBuilder: PlaybackStateCompat.Builder
    private lateinit var mCallbacks: SaundioMediaCallback

    override fun onCreate() {
        super.onCreate()

        mMediaSession = MediaSessionCompat(applicationContext, MediaService::class.java.simpleName)
        mMediaSession.setFlags(MediaSessionCompat.FLAG_HANDLES_MEDIA_BUTTONS or MediaSessionCompat.FLAG_HANDLES_TRANSPORT_CONTROLS)

        mStateBuilder = PlaybackStateCompat.Builder().setActions(POSSIBLE_STATE_FLAGS)
        mMediaSession.setPlaybackState(mStateBuilder.build())

        mCallbacks = SaundioMediaCallback(this, mMediaSession, mStateBuilder)
        mMediaSession.setCallback(mCallbacks)

        sessionToken = mMediaSession.sessionToken
    }

    override fun onDestroy() {
        mMediaSession.release()
        super.onDestroy()
    }

    override fun onGetRoot(clientPackageName: String, clientUid: Int, rootHints: Bundle?): MediaBrowserServiceCompat.BrowserRoot? {
        return if (TextUtils.equals(clientPackageName, packageName)) MediaBrowserServiceCompat.BrowserRoot(getString(R.string.app_name), null) else null

    }

    override fun onLoadChildren(parentId: String, result: MediaBrowserServiceCompat.Result<List<MediaBrowserCompat.MediaItem>>) {
        result.sendResult(null)
    }

    companion object {

        private const val POSSIBLE_STATE_FLAGS = ACTION_PLAY or
                ACTION_PAUSE or
                ACTION_PLAY_PAUSE or
                ACTION_PLAY_FROM_MEDIA_ID or
                ACTION_PLAY_FROM_URI or
                ACTION_PLAY_FROM_SEARCH or
                ACTION_FAST_FORWARD or
                ACTION_REWIND or
                ACTION_SKIP_TO_NEXT or
                ACTION_SKIP_TO_PREVIOUS or
                ACTION_SEEK_TO
    }
}
